mmtk/src/collection.rs

```diff
@@ -8,11 +8,13 @@ use crate::{MutatorClosure, OpenJDK};
 
 pub struct VMCollection {}
 
-extern "C" fn report_mutator_stop<F>(mutator: *mut Mutator<OpenJDK>, callback: *mut F)
-where
+extern "C" fn report_mutator_stop<F>(
+    mutator: *mut Mutator<OpenJDK>,
+    callback_ptr: *mut libc::c_void,
```

Comment:
Why is this changed from `*mut F` to `*mut c_void`?

Reply:
Because the `MutatorClosure` struct needs to be passed to C.

```
#[repr(C)]
pub struct MutatorClosure {
    pub func: extern "C" fn(mutator: *mut Mutator<OpenJDK>, data: *mut libc::c_void),
    pub data: *mut libc::c_void,
}
```

The contract here is, C/C++ programs treat `data` as an opaque pointer, and use it as the last argument of `func`.  The `data` may not point to the same thing each time, so it is `*mut libc::c_void`.

Another reason is, Rust forbids the cast between function pointer types with different parameter types, so the line `func: report_mutator_stop::<F> as *const _` which used to work now stopped working.

So the only solution is to let the Rust function `report_mutator_stop<F>` match the parameter type `*mut libc::c_void`.